### PR TITLE
Move some CSI components to Seed cluster

### PR DESCRIPTION
**Pull request by jia-jerry**
**What this PR does / why we need it**:
Move some CSI components to Seed cluster. 
Fix bug when shoot cluster is hibernized and has some PV in the cluster, user can't delete it successfully.

### CODE REVIEW

File: charts/seed-controlplane/charts/csi-alicloud/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+description: Helm chart for kubernetes cloud-metadata-service
+name: cloud-metadata-service
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/Chart.yaml]: Wrong name, wrong description

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-plugin-dep.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  template:
+    metadata:
+      annotations:
+        checksum/secret-csi-diskplugin-alicloud: {{ include (print $.Template.BasePath "/csi-diskplugin-secret.yaml") . | sha256sum }}
+      labels:
+        garden.sapcloud.io/role: controlplane
+        app: kubernetes
+        role: csi-plugin-controller
+    spec:
+      containers:
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-plugin-dep.yaml:26]: Specify some resource requests/limits.

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-attacher-dep.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      app: kubernetes
+      role: csi-attacher
+  template:
+    metadata:
+      labels:
+        app: kubernetes
+        role: csi-attacher
+        garden.sapcloud.io/role: controlplane
+    spec:
+      containers:
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-attacher-dep.yaml:27]: Specify some resource requests/limits.

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      app: kubernetes
+      role: csi-provisioner
+  template:
+    metadata:
+      labels:
+        app: kubernetes
+        role: csi-provisioner
+        garden.sapcloud.io/role: controlplane
+    spec:
+      containers:
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml:27]: Specify some resource requests/limits.

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    role: csi-snapshotter
+    garden.sapcloud.io/role: controlplane
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: kubernetes
+      role: csi-snapshotter
+  template:
+    metadata:
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml:17]: secret checksum annotation missing

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      labels:
+        app: kubernetes
+        role: csi-snapshotter
+        garden.sapcloud.io/role: controlplane
+    spec:
+      containers:
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml:27]: Specify some resource requests/limits.

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    role: csi-provisioner
+    garden.sapcloud.io/role: controlplane
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: kubernetes
+      role: csi-provisioner
+  template:
+    metadata:
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml:17]: secret checksum annotation missing

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-diskplugin-secret.yaml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-diskplugin-secret.yaml]: Can't you re-use the secret with name `cloudprovider`? It already has all the information you require?

File: pkg/operation/botanist/controlplane.go [hunk 9]
```diff
@@ -627,7 +627,11 @@ func (b *Botanist) WakeUpControlPlane(ctx context.Context) error {
 		return err
 	}
 
-	for _, deployment := range []string{common.KubeControllerManagerDeploymentName, common.CloudControllerManagerDeploymentName, common.MachineControllerManagerDeploymentName} {
+	deployments := []string{common.KubeControllerManagerDeploymentName, common.CloudControllerManagerDeploymentName, common.MachineControllerManagerDeploymentName}
+	if b.Shoot.IsHibernated {
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
	if b.Shoot.UsesCSI() {
```

File: pkg/operation/botanist/controlplane.go [hunk 10]
```diff
@@ -651,7 +655,11 @@ func (b *Botanist) HibernateControlPlane(ctx context.Context) error {
 		}
 	}
 
-	for _, deployment := range []string{common.KubeAddonManagerDeploymentName, common.KubeControllerManagerDeploymentName, common.KubeAPIServerDeploymentName} {
+	deployments := []string{common.KubeAddonManagerDeploymentName, common.KubeControllerManagerDeploymentName, common.KubeAPIServerDeploymentName}
+	if b.Shoot.IsHibernated {
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
	if b.Shoot.UsesCSI() {
```

File: pkg/operation/cloudbotanist/alicloudbotanist/addons.go [hunk 11]
```diff
@@ -44,6 +46,27 @@ func (b *AlicloudBotanist) GenerateNginxIngressConfig() (map[string]interface{},
...
 
+// GenerateCSIConfig generates the configuration for CSI charts
+func (b *AlicloudBotanist) GenerateCSIConfig() (map[string]interface{}, error) {
+	conf := map[string]interface{}{
+		"credential": map[string]interface{}{
+			"accessKeyID":     base64.StdEncoding.EncodeToString(b.Shoot.Secret.Data[AccessKeyID]),
+			"accessKeySecret": base64.StdEncoding.EncodeToString(b.Shoot.Secret.Data[AccessKeySecret]),
+		},
+		"regionID":          b.Shoot.Info.Spec.Cloud.Region,
+		"kubernetesVersion": b.Operation.ShootVersion(),
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/alicloudbotanist/addons.go]: This is not used anywhere - remove it.

File: pkg/operation/cloudbotanist/alicloudbotanist/addons.go [hunk 12]
```diff
@@ -44,6 +46,27 @@ func (b *AlicloudBotanist) GenerateNginxIngressConfig() (map[string]interface{},
 	}, b.Shoot.NginxIngressEnabled()), nil
 }
 
+// GenerateCSIConfig generates the configuration for CSI charts
+func (b *AlicloudBotanist) GenerateCSIConfig() (map[string]interface{}, error) {
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/alicloudbotanist/addons.go]: Can you move this into `controlplane.go`?

File: pkg/operation/hybridbotanist/addons.go [hunk 13]
```diff
@@ -148,11 +148,6 @@ func (b *HybridBotanist) generateCoreAddonsChart() (*chartrenderer.RenderedChart
 		return nil, err
 	}
 
-	ccmConfig := map[string]interface{}{}
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/addons.go:151]: Can you also remove https://github.com/gardener/gardener/blob/master/charts/shoot-core/charts/cloud-controller-manager/values.yaml#L1?

File: pkg/operation/hybridbotanist/addons.go [hunk 14]
```diff
@@ -167,7 +162,7 @@ func (b *HybridBotanist) generateCoreAddonsChart() (*chartrenderer.RenderedChart
 			"node-exporter":     nodeExporter,
 			"blackbox-exporter": blackboxExporter,
 		},
-		"cloud-controller-manager": ccmConfig,
+		"cloud-controller-manager": map[string]interface{}{},
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/addons.go]: Can you simply completely remove this line?

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by rfranzke** [comment]
FYI: I changed the base branch to `master`. Please rebase accordingly.
In general, first create PRs for `master`. If they are successfully merged then you can create another PR for the respective release/hotfix branch. :-)

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml (hunk 5 above)
Comment (jia-jerry) [reply charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml:17]: No secret needed for snapshotter.

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml (hunk 7 above)
Comment (jia-jerry) [reply charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml:17]: No secret needed for provisioner

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-diskplugin-secret.yaml (hunk 8 above)
Comment (jia-jerry) [reply charts/seed-controlplane/charts/csi-alicloud/templates/csi-diskplugin-secret.yaml]: Yes. Thanks for your suggestion.

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml (hunk 5 above)
Comment (rfranzke) [reply charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml:17]: You are mounting secret "csi-snapshotter"

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml (hunk 7 above)
Comment (rfranzke) [reply charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml:17]: You are mounting secret "csi-snapshotter"

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml (hunk 5 above)
Comment (jia-jerry) [reply charts/seed-controlplane/charts/csi-alicloud/templates/csi-snapshotter-dep.yaml:17]: Oops. sorry.

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-plugin-dep.yaml [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+{{- end }}     
+      labels:
+        garden.sapcloud.io/role: controlplane
+        app: kubernetes
+        role: csi-plugin-controller
+    spec:
+      containers:
+        - name: csi-diskplugin
+          image: {{ index .Values.images "csi-plugin-alicloud" }}
+          imagePullPolicy: Always
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-plugin-dep.yaml]:
```suggestion
          imagePullPolicy: IfNotPresent
```

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+            - "--kubeconfig=/var/lib/csi-provisioner/kubeconfig"
+            - "--feature-gates=Topology=True"
+            - "--enable-leader-election=true"
+{{- if .Values.provisionerResources }}
+          resources:
+{{ toYaml .Values.provisionerResources | indent 12 }}           
+{{- end }}
+          env:
+          - name: POD_NAMESPACE
+            value: kube-system
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml:42]: Is this still correct?

File: pkg/operation/cloudbotanist/alicloudbotanist/controlplane.go [hunk 17]
```diff
@@ -165,6 +145,40 @@ func (b *AlicloudBotanist) GenerateKubeAPIServerExposeConfig() (map[string]inter
...
+		},
+		"enabled": true,
+	}
+
+	return b.ImageVector.InjectImages(conf, b.ShootVersion(), b.ShootVersion(),
+		common.CSIAttacherImageName,
+		common.CSIPluginAlicloudImageName,
+		common.CSIProvisionerImageName,
+		common.CSISnapshotterImageName,
+		common.CSINodeDriverRegistrarImageName,
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/alicloudbotanist/controlplane.go:178]: This is not needed here, right?

File: pkg/operation/cloudbotanist/alicloudbotanist/controlplane.go [hunk 18]
```diff
@@ -165,6 +145,40 @@ func (b *AlicloudBotanist) GenerateKubeAPIServerExposeConfig() (map[string]inter
...
+			"csiSnapshotter": map[string]interface{}{
+				fmt.Sprintf("checksum/%s", common.CSISnapshotter): b.CheckSums[common.CSISnapshotter],
+			},
+		},
+		"enabled": true,
+	}
+
+	return b.ImageVector.InjectImages(conf, b.ShootVersion(), b.ShootVersion(),
+		common.CSIAttacherImageName,
+		common.CSIPluginAlicloudImageName,
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/alicloudbotanist/controlplane.go:175]: This is not needed here, right?

File: pkg/operation/cloudbotanist/alicloudbotanist/controlplane.go (hunk 17 above)
Comment (jia-jerry) [reply pkg/operation/cloudbotanist/alicloudbotanist/controlplane.go:178]: It is needed. function GenerateCSIConfig() is also used by csi-alicloud in **shoot-core** charts

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml (hunk 16 above)
Comment (jia-jerry) [reply charts/seed-controlplane/charts/csi-alicloud/templates/csi-provisioner-dep.yaml:42]: We passed shoot api server config file, so we use kube-system namespace to create endpoint lock for provisioner.

File: pkg/operation/cloudbotanist/alicloudbotanist/controlplane.go (hunk 18 above)
Comment (jia-jerry) [reply pkg/operation/cloudbotanist/alicloudbotanist/controlplane.go:175]: We still need it. This function is used for 2 charts.

File: charts/seed-controlplane/charts/csi-alicloud/templates/csi-plugin-dep.yaml (hunk 15 above)
Comment (jia-jerry) [reply charts/seed-controlplane/charts/csi-alicloud/templates/csi-plugin-dep.yaml]: done.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**